File: app/src/main/res/layout/ui/admin/StatisticalLecturerActivity.java
package com.fh.app_student_management.ui.admin;

import android.content.DialogInterface;
import android.os.Bundle;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.fh.app_student_management.R;
import com.fh.app_student_management.adapters.admin.LecturerStatisticalRecycleViewAdapter;
import com.fh.app_student_management.data.AppDatabase;
import com.fh.app_student_management.data.entities.Semester;
import com.fh.app_student_management.data.relations.LecturerAndUser;
import com.fh.app_student_management.data.relations.StatisticalOfLecturer;
import com.fh.app_student_management.utilities.Utils;

import java.util.ArrayList;
import java.util.List;

public class StatisticalLecturerActivity extends AppCompatActivity {

    private ArrayList<Semester> semesters;
    private ArrayList<String> semesterNames;
    private long selectedSemesterId;
    private ArrayList<LecturerAndUser> lecturers;
    private ArrayList<String> lecturerNames;
    private long selectedLecturerId;

    private ImageView btnBack;
    private EditText edtSemester;
    private EditText edtLecturer;
    private LinearLayout titleTable;
    private TextView txtLecturerName;
    private TextView txtSemesterName;
    private RecyclerView rvLecturer;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.admin_activity_statistical_lecturer);
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });

        initStatisticalLecturerView();
        handleEventListener();
    }

    private void initStatisticalLecturerView() {
        btnBack = findViewById(R.id.btnBack);
        edtSemester = findViewById(R.id.edtSemester);
        edtLecturer = findViewById(R.id.edtLecturer);
        titleTable = findViewById(R.id.titleTable);
        txtLecturerName = findViewById(R.id.txtLecturerName);
        txtSemesterName = findViewById(R.id.txtSemesterName);
        rvLecturer = findViewById(R.id.rvLecturer);

        titleTable.setVisibility(View.GONE);

        semesters = new ArrayList<>(AppDatabase.getInstance(this).semesterDAO().getAll());
        semesterNames = new ArrayList<>(semesters.size() + 1);
        semesterNames.add(0, "--- Chọn học kỳ ---");
        for (int i = 0; i < semesters.size(); i++) {
            String startDate = Utils.formatDate("MM/yyyy").format(semesters.get(i).getStartDate());
            String endDate = Utils.formatDate("MM/yyyy").format(semesters.get(i).getEndDate());
            String semesterName = String.format("%s (%s - %s)", semesters.get(i).getName(), startDate, endDate);
            semesterNames.add(semesterName);
        }
    }

    private void handleEventListener() {
        btnBack.setOnClickListener(v -> finish());

        edtSemester.setOnClickListener(v -> showSelectionDialog("Chọn học kỳ", semesterNames, (dialog, which) -> {
            if (which == 0) {
                resetSelections(edtSemester, edtLecturer);
            } else {
                selectedSemesterId = semesters.get(which - 1).getId();
                edtSemester.setText(semesterNames.get(which));
                resetSelections(edtLecturer);
            }
        }));

        edtLecturer.setOnClickListener(v -> {
            if (edtSemester.getText().toString().isEmpty()) {
                Utils.showToast(this, "Chưa chọn học kỳ");
                return;
            }

            lecturers = new ArrayList<>(AppDatabase.getInstance(this)
                    .lecturerDAO().getAllLecturerAndUserBySemester(selectedSemesterId));
            lecturerNames = new ArrayList<>(lecturers.size() + 1);
            lecturerNames.add(0, "--- Chọn giảng viên ---");
            for (int i = 0; i < lecturers.size(); i++) {
                lecturerNames.add(lecturers.get(i).getUser().getFullName());
            }

            showSelectionDialog("Chọn giảng viên", lecturerNames, (dialog, which) -> {
                if (which == 0) {
                    resetSelections(edtLecturer);
                } else {
                    selectedLecturerId = lecturers.get(which - 1).getLecturer().getId();
                    edtLecturer.setText(lecturerNames.get(which));
                    titleTable.setVisibility(View.VISIBLE);
                    txtLecturerName.setText(lecturerNames.get(which));
                    txtSemesterName.setText(edtSemester.getText().toString());
                    updateStatistical();
                }
            });
        });
    }

    private void updateStatistical() {
        List<StatisticalOfLecturer> statisticalOfLecturers = new ArrayList<>();
        if (selectedSemesterId > 0 && selectedLecturerId > 0) {
            statisticalOfLecturers = AppDatabase.getInstance(this)
                    .statisticalDAO().getStatisticalOfLecturer(selectedSemesterId, selectedLecturerId);
        }
        rvLecturer.setLayoutManager(new LinearLayoutManager(this));
        rvLecturer.setAdapter(new LecturerStatisticalRecycleViewAdapter(this, new ArrayList<>(statisticalOfLecturers)));
    }

    private void showSelectionDialog(String title, @NonNull List<String> options, DialogInterface.OnClickListener listener) {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(title);
        builder.setItems(options.toArray(new CharSequence[0]), listener);
        builder.show();
    }

    private void resetSelections(@NonNull EditText edtSemester, EditText edtLecturer) {
        titleTable.setVisibility(View.GONE);
        selectedSemesterId = 0;
        edtSemester.setText("");
        resetSelections(edtLecturer);
    }

    private void resetSelections(@NonNull EditText edtLecturer) {
        titleTable.setVisibility(View.GONE);
        selectedLecturerId = 0;
        edtLecturer.setText("");
        lecturerNames = null;
        updateStatistical();
    }
}

File: app/src/main/java/ma/ensa/projet/ui/admin/StatisticalLecturerActivity.kt
package ma.ensa.projet.ui.admin


import android.content.DialogInterface
import android.os.Bundle
import android.view.View
import android.widget.EditText
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.graphics.Insets
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import ma.ensa.projet.R
import ma.ensa.projet.adapters.admin.LecturerStatisticalRecycleViewAdapter
import ma.ensa.projet.data.AppDatabase
import ma.ensa.projet.data.dto.LecturerAndUser
import ma.ensa.projet.data.dto.StatisticalOfLecturer
import ma.ensa.projet.data.entities.Semester
import ma.ensa.projet.utilities.Utils

class StatisticalLecturerActivity : AppCompatActivity() {

    private lateinit var semesters: ArrayList<Semester>
    private lateinit var semesterNames: ArrayList<String>
    private var selectedSemesterId: Long = 0
    private lateinit var lecturers: ArrayList<LecturerAndUser>
    private lateinit var lecturerNames: ArrayList<String>
    private var selectedLecturerId: Long = 0

    private lateinit var btnBack: ImageView
    private lateinit var edtSemester: EditText
    private lateinit var edtLecturer: EditText
    private lateinit var titleTable: LinearLayout
    private lateinit var txtLecturerName: TextView
    private lateinit var txtSemesterName: TextView
    private lateinit var rvLecturer: RecyclerView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.admin_activity_statistical_lecturer)

        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        initStatisticalLecturerView()
        handleEventListener()
    }

    private fun initStatisticalLecturerView() {
        btnBack = findViewById(R.id.btnBack)
        edtSemester = findViewById(R.id.edtSemester)
        edtLecturer = findViewById(R.id.edtLecturer)
        titleTable = findViewById(R.id.titleTable)
        txtLecturerName = findViewById(R.id.txtLecturerName)
        txtSemesterName = findViewById(R.id.txtSemesterName)
        rvLecturer = findViewById(R.id.rvLecturer)

        titleTable.visibility = View.GONE

        // Launch a coroutine to fetch semesters
        lifecycleScope.launch(Dispatchers.IO) {
            semesters = ArrayList(AppDatabase.getInstance(this@StatisticalLecturerActivity)?.semesterDAO()?.getAll() ?: emptyList())
            val semesterNamesList = mutableListOf<String>().apply {
                add("--- Select Semester ---")  // Translation of "Chọn học kỳ"
                for (semester in semesters) {
                    val startDate = Utils.formatDate("MM/yyyy").format(semester.startDate)
                    val endDate = Utils.formatDate("MM/yyyy").format(semester.endDate)
                    add("${semester.name} ($startDate - $endDate)")
                }
            }

            // Switch to the main thread to update the UI
            withContext(Dispatchers.Main) {
                semesterNames = ArrayList(semesterNamesList)
            }
        }
    }

    private fun handleEventListener() {
        btnBack.setOnClickListener { finish() }

        edtSemester.setOnClickListener {
            showSelectionDialog("Select Semester", semesterNames) { dialog, which ->
                if (which == 0) {
                    resetSelections(edtSemester, edtLecturer)
                } else {
                    selectedSemesterId = semesters[which - 1].id
                    edtSemester.setText(semesterNames[which])
                    resetSelections(edtLecturer)
                }
            }
        }

        edtLecturer.setOnClickListener {
            if (edtSemester.text.toString().isEmpty()) {
                Utils.showToast(this, "Semester not selected")
                return@setOnClickListener
            }

            // Launch a coroutine to fetch lecturers based on selected semester
            lifecycleScope.launch(Dispatchers.IO) {
                lecturers = ArrayList(AppDatabase.getInstance(this@StatisticalLecturerActivity)?.lecturerDAO()?.getAllLecturerAndUserBySemester(selectedSemesterId) ?: emptyList())
                val lecturerNamesList = mutableListOf<String>().apply {
                    add("--- Select Lecturer ---")
                    for (lecturer in lecturers) {
                        lecturer.user.fullName?.let { add(it) }
                    }
                }

                // Show the selection dialog on the main thread
                withContext(Dispatchers.Main) {
                    showSelectionDialog("Select Lecturer", ArrayList(lecturerNamesList)) { dialog, which ->
                        if (which == 0) {
                            resetSelections(edtLecturer)
                        } else {
                            selectedLecturerId = lecturers[which - 1].lecturer.id
                            edtLecturer.setText(lecturerNamesList[which])
                            titleTable.visibility = View.VISIBLE
                            txtLecturerName.text = lecturerNamesList[which]
                            txtSemesterName.text = edtSemester.text.toString()
                            updateStatistical()
                        }
                    }
                }
            }
        }
    }

    private fun updateStatistical() {
        lifecycleScope.launch(Dispatchers.IO) {
            val statisticalOfLecturers = if (selectedSemesterId > 0 && selectedLecturerId > 0) {
                AppDatabase.getInstance(this@StatisticalLecturerActivity)
                    ?.statisticalDAO()?.getStatisticalOfLecturer(selectedSemesterId, selectedLecturerId) ?: emptyList()
            } else {
                emptyList<StatisticalOfLecturer>()
            }

            // Switch to the main thread to update the UI
            withContext(Dispatchers.Main) {
                rvLecturer.layoutManager = LinearLayoutManager(this@StatisticalLecturerActivity)
                rvLecturer.adapter = LecturerStatisticalRecycleViewAdapter(this@StatisticalLecturerActivity, ArrayList(statisticalOfLecturers))
            }
        }
    }


    private fun showSelectionDialog(title: String, options: List<String>, listener: DialogInterface.OnClickListener) {
        val builder = AlertDialog.Builder(this)
        builder.setTitle(title)
        builder.setItems(options.toTypedArray(), listener)
        builder.show()
    }

    private fun resetSelections(edtSemester: EditText, edtLecturer: EditText) {
        titleTable.visibility = View.GONE
        selectedSemesterId = 0
        edtSemester.setText("")
        resetSelections(edtLecturer)
    }

    private fun resetSelections(edtLecturer: EditText) {
        titleTable.visibility = View.GONE
        selectedLecturerId = 0
        edtLecturer.setText("")
        lecturerNames = ArrayList()
        updateStatistical()
    }
}
